### ISSUE TITLE
Kernel page fault when running qemu

### ISSUE THREAD

**Issue by qookei** [managarm/managarm#407]
As mentioned in managarm/bootstrap-managarm#151, running qemu when running with more than one core causes a page fault in the IPI handling logic (internal linked list of the `async::shared_mutex` is seemingly empty due to `_front == nullptr`).

Backtrace:
```
(gdb) bt
#0  thor::panic () at ../../../src/managarm/kernel/thor/generic/debug.cpp:149
#1  thor::PanicSink::operator() (this=<optimized out>, msg=<optimized out>) at ../../../src/managarm/kernel/thor/generic/debug.cpp:187
#2  0xffffffff80133033 in frg::stack_buffer_logger<thor::PanicSink, 128ul>::_emit (this=0xffffffff80190950 <thor::pioLogHandler>, 
    message=0xffffe00000c0fc20 "Fault #14, with IF=0, cs: 0x20, ip: 0xffffffff80012af5\nError code: 0x0, SS: 0x38, RSP: 0xffffe00000c0fdb0")
    at ../../../src/managarm/subprojects/frigg/include/frg/logging.hpp:92
#3  frg::stack_buffer_logger<thor::PanicSink, 128ul>::item::operator<< (this=0xffffe00000c0fc18) at ../../../src/managarm/subprojects/frigg/include/frg/logging.hpp:44
#4  onPlatformFault (image=..., number=<optimized out>) at ../../../src/managarm/kernel/thor/arch/x86/ints.cpp:333
#5  0xffffffff80000518 in faultStubPage () at ../../../src/managarm/kernel/thor/arch/x86/stubs.S:200
#6  0xffffffff80012af5 in frg::_list::intrusive_list<async::detail::shared_mutex::node, frg::locate_member<async::detail::shared_mutex::node, frg::_list::intrusive_list_hook<async::detail::shared_mutex::node*, async::detail::shared_mutex::node*>, &async::detail::shared_mutex::node::hook> >::front (this=0xffffe00007621c70)
    at ../../../src/managarm/subprojects/frigg/include/frg/list.hpp:148
#7  async::detail::shared_mutex::unlock (this=0xffffe00007621c60) at ../../../src/managarm/subprojects/libasync/include/async/mutex.hpp:313
#8  0xffffffff80009148 in frg::unique_lock<async::detail::shared_mutex>::unlock (this=<optimized out>) at ../../../src/managarm/subprojects/frigg/include/frg/mutex.hpp:63
#9  frg::unique_lock<async::detail::shared_mutex>::~unique_lock (this=<optimized out>) at ../../../src/managarm/subprojects/frigg/include/frg/mutex.hpp:47
#10 thor::VirtualSpace::unmap (this=<optimized out>, address=<optimized out>, length=<optimized out>) at ../../../src/managarm/kernel/thor/generic/address-space.cpp:582
#11 0xffffffff8013d3e6 in thor::PageBinding::shootdown (this=0xffffe00000a37128) at ../../../src/managarm/kernel/thor/arch/x86/paging.cpp:317
#12 0xffffffff801355a4 in onPlatformShootdown (image=...) at ../../../src/managarm/kernel/thor/arch/x86/ints.cpp:459
#13 0xffffffff80001ec7 in thorRtIpiShootdown () at ../../../src/managarm/kernel/thor/arch/x86/stubs.S:445
#14 0x0000000000200046 in ?? ()
#15 0xffffe00000c50000 in ?? ()
#16 0xffffe0000648c101 in ?? ()
#17 0xffffe00000c3ff4e in ?? ()
#18 0xffffe00000c3ff4e in ?? ()
#19 0xffffe00000c3ff20 in ?? ()
#20 0x94d049bb133111eb in ?? ()
#21 0x000000002c6ac000 in ?? ()
#22 0x00007fffffa3808c in ?? ()
#23 0xffffe0000591faf0 in ?? ()
#24 0xffffe0000648c018 in ?? ()
#25 0x000000008000001f in ?? ()
#26 0xffffffff800d7fc0 in ?? () at ../../../src/managarm/kernel/thor/arch/x86/thor-internal/arch/cpu.hpp:627
#27 0xffffe00000c3ff4e in ?? ()
#28 0xffffe00000c4ffe0 in ?? ()
#29 0xffffffff8012954c in halt_loop () at ../../../src/managarm/kernel/thor/arch/x86/entry.S:36
#30 0x0000000000000058 in ?? ()
#31 0xffffffff800d7fc9 in frg::_tuple::apply<thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, thor::Continuation, 0ul>(thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, frg::tuple<thor::Continuation>&&, std::integer_sequence<unsigned long, 0ul>) (functor=..., args=...)
    at ../../../src/managarm/subprojects/frigg/include/frg/tuple.hpp:146
#32 frg::apply<thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, thor::Continuation>(thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, frg::tuple<thor::Continuation>&&) (args=..., functor=...) at ../../../src/managarm/subprojects/frigg/include/frg/tuple.hpp:246
#33 thor::runOnStack<thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}>(thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, thor::StackBase)::{lambda(void*, void*)#1}::operator()(void*, void*) const (this=<optimized out>, context=0xffffe00000c0fdb0, previousSp=0x0)
    at ../../../src/managarm/kernel/thor/arch/x86/thor-internal/arch/cpu.hpp:624
#34 thor::runOnStack<thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}>(thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}, thor::StackBase)::{lambda(void*, void*)#1}::__invoke(void*, void*) (context=0xffffe00000c0fdb0, previousSp=0x0) at ../../../src/managarm/kernel/thor/arch/x86/thor-internal/arch/cpu.hpp:622
--Type <RET> for more, q to quit, c to continue without paging--c
#35 0xffffffff8012aa9a in thor::doRunOnStack (function=0x58, sp=0xffffffff8012954a <enter_context>, argument=0xffffffff800d7fd9 <thor::(anonymous namespace)::IdleTask::invoke()::{lambda(thor::Continuation)#1}::operator()(thor::Continuation) const+9>) at ../../../src/managarm/kernel/thor/arch/x86/cpu.cpp:517
```

**Comment by qookei** [comment]
Presumably fixed by managarm/libasync@d8983f79e76f1756ac1a677c86a7880edb99bf27? Need to check first before closing.

**Comment by qookei** [comment]
Can't reproduce, seems like it's fixed? Closing
